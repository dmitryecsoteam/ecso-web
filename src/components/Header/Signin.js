import React from 'react';
import { Mutation } from 'react-apollo';

import InputWithErrorTooltip from '../MainPage/InputWithErrorTooltip';
import { SIGNIN_USER } from '../../queries/mutations';


export default class Signin extends React.Component {
    state = {
        email: '',
        password: '',
        emailError: false,
        emailErrorText: '',
        passwordError: false,
        passwordErrorText: ''
    }

    onInputChange = (event) => {
        const { name, value } = event.target;

        this.setState({
            [name]: value
        });
    }

    onFormSubmit = async (event, signinUser) => {
        event.preventDefault();

        let emailError = false,
            emailErrorText = '',
            passwordError = false,
            passwordErrorText = '';

            const { email, password } = this.state;

        if (!email) {
            emailError = true;
            emailErrorText = 'Enter your email';
        }

        if (!password) {
            passwordError = true;
            passwordErrorText = 'Enter password';
        }

        this.setState({
            emailError,
            emailErrorText,
            passwordError,
            passwordErrorText
        });

        if (!(emailError || passwordError)) {
            try {

                const data = await signinUser();
                console.log(data)

            } catch (e) {

                if (e.message.includes('User with email')) {
                    this.setState({
                        emailError: true,
                        emailErrorText: e.message.replace('GraphQL error: ', '')
                    });
                }

                if (e.message.includes('Wrong password')) {
                    this.setState({
                        passwordError: true,
                        passwordErrorText: e.message.replace('GraphQL error: ', '')
                    });
                }
            }
        }
    }

    render() {

        const {
            email,
            password,
            emailError,
            emailErrorText,
            passwordError,
            passwordErrorText
        } = this.state;

        return (
            <div>
                <h2 className="signup__title">Signin with email</h2>

                <Mutation mutation={SIGNIN_USER} variables={{ email, password }}>

                    {(signinUser, { data, loading, error }) => {

                        return (
                            <form
                                onSubmit={e => this.onFormSubmit(e, signinUser)}
                                autoComplete="off"
                            >

                                <InputWithErrorTooltip
                                    error={emailError}
                                    errorText={emailErrorText}
                                >
                                    <input
                                        type="text"
                                        name="email"
                                        value={email}
                                        placeholder="Email"
                                        className="signup__input"
                                        onChange={this.onInputChange}
                                    />
                                </InputWithErrorTooltip>

                                <InputWithErrorTooltip
                                    error={passwordError}
                                    errorText={passwordErrorText}
                                >
                                    <input
                                        type="password"
                                        name="password"
                                        value={password}
                                        placeholder="Password"
                                        className="signup__input"
                                        onChange={this.onInputChange}
                                    />
                                </InputWithErrorTooltip>

                                <button className="search-form__submit-btn">Signin</button>
                            </form>
                        )
                    }}

                </Mutation>
            </div>
        );
    }
}