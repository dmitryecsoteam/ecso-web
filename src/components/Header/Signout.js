import React from 'react';
import { ApolloConsumer } from 'react-apollo';
import { withRouter } from "react-router-dom";

const handleSignout = (client, history) => {
    console.log('Loggin out')
    localStorage.setItem("token", "");
    client.resetStore();
    history.push('/');
}

export const Signout = ({ history }) => (
    <ApolloConsumer>
        {client => {
            return <div 
                    onClick={() => handleSignout(client, history)}
                    className="user-menu__link"
                >
                    Log out
                </div>
        }}
    </ApolloConsumer>
);

export default withRouter(Signout);
