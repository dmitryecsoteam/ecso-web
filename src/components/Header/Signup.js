import React from 'react';
import { Mutation } from 'react-apollo';
import SHA256 from 'crypto-js/sha256';

import InputWithErrorTooltip from '../MainPage/InputWithErrorTooltip';
import { SIGNUP_USER } from '../../queries/mutations';


export default class Signup extends React.Component {
  state = {
    email: '',
    name: '',
    password: '',
    passwordConfirmation: '',
    emailError: false,
    emailErrorText: '',
    nameError: false,
    nameErrorText: '',
    passwordError: false,
    passwordErrorText: '',
    passwordConfirmationError: false,
    passwordConfirmationErrorText: '',
  }

  onInputChange = (event) => {
    const { name, value } = event.target;

    this.setState({
      [name]: value
    });
  }

  onFormSubmit = async (event, signupUser) => {
    event.preventDefault();

    const { email, name, password, passwordConfirmation } = this.state;

    let emailError = false,
      emailErrorText = '',
      nameError = false,
      nameErrorText = '',
      passwordError = false,
      passwordErrorText = '',
      passwordConfirmationError = false,
      passwordConfirmationErrorText = '';

    if (!email) {
      emailError = true;
      emailErrorText = 'Enter your email';
    }

    if (!name) {
      nameError = true;
      nameErrorText = 'Enter your name';
    }

    if (!password) {
      passwordError = true;
      passwordErrorText = 'Enter password';
    }

    if (!passwordConfirmation) {
      passwordConfirmationError = true;
      passwordConfirmationErrorText = 'Confirm your password'
    } else {
      if (password !== passwordConfirmation) {
        passwordConfirmationError = true;
        passwordConfirmationErrorText = 'Passwords didn\'t match';
      }
    }

    this.setState({
      emailError,
      emailErrorText,
      nameError,
      nameErrorText,
      passwordError,
      passwordErrorText,
      passwordConfirmationError,
      passwordConfirmationErrorText,
    });



    if (!(emailError || nameError || passwordError || passwordConfirmationError)) {
      try {

        const { data } = await signupUser();
        console.log(data);

        localStorage.setItem('token', data.signupUser.token);

        this.props.fetchUser();

        if (this.props.closeModal) this.props.closeModal();

      } catch (e) {

        console.log(e)
        this.setState({
          emailError: true,
          emailErrorText: e.message.replace('GraphQL error: ', '')
        });

      }
    }


  }

  render() {

    const {
      email,
      name,
      password,
      passwordConfirmation,
      emailError,
      emailErrorText,
      nameError,
      nameErrorText,
      passwordError,
      passwordErrorText,
      passwordConfirmationError,
      passwordConfirmationErrorText,
    } = this.state;

    const passwordSHA = SHA256(password).toString();

    return (
      <div>
        <h2 className="signup__title">Signup with email</h2>

        <Mutation mutation={SIGNUP_USER} variables={{ email, password: passwordSHA, name }}>

          {(signupUser, { data, loading, error }) => {

            return (
              <form
                onSubmit={e => this.onFormSubmit(e, signupUser)}
                autoComplete="off"
              >

                <InputWithErrorTooltip
                  error={emailError}
                  errorText={emailErrorText}
                >
                  <input
                    type="text"
                    name="email"
                    value={email}
                    placeholder="Email"
                    className="signup__input"
                    onChange={this.onInputChange}
                  />
                </InputWithErrorTooltip>

                <InputWithErrorTooltip
                  error={nameError}
                  errorText={nameErrorText}
                >
                  <input
                    type="text"
                    name="name"
                    value={name}
                    placeholder="Name"
                    className="signup__input"
                    onChange={this.onInputChange}
                  />
                </InputWithErrorTooltip>

                <InputWithErrorTooltip
                  error={passwordError}
                  errorText={passwordErrorText}
                >
                  <input
                    type="password"
                    name="password"
                    value={password}
                    placeholder="Password"
                    className="signup__input"
                    onChange={this.onInputChange}
                  />
                </InputWithErrorTooltip>

                <InputWithErrorTooltip
                  error={passwordConfirmationError}
                  errorText={passwordConfirmationErrorText}
                >
                  <input
                    type="password"
                    name="passwordConfirmation"
                    value={passwordConfirmation}
                    placeholder="Confirm password"
                    className="signup__input"
                    onChange={this.onInputChange}
                  />
                </InputWithErrorTooltip>

                <button className="search-form__submit-btn">Submit</button>
              </form>);
          }}

        </Mutation>

      </div>
    );
  }
}