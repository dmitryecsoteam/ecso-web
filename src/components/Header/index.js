import React from 'react';
import { Link } from 'react-router-dom';

import logo from '../../images/logo/logo2-darkgrey.png';

export default () => (
    <header className="header">
        <div className="header__logo-container">
            <Link to="/">
                <img className="header__image" src={logo} alt={"logo"} />
            </Link>
        </div>

        <nav>
            <ul className="header__menu-container">
                <li className="header__menu-item">
                    <Link className="header__menu-link" to="/help">Help</Link>
                </li>
                <li className="header__menu-item">
                    <Link className="header__menu-link" to="/signin">Signin</Link>
                </li>
                <li className="header__menu-item">
                    <Link className="header__menu-link" to="/signup">Signup</Link>
                </li>
            </ul>
        </nav>
    </header>
);