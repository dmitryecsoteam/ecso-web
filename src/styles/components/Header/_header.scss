.header {
  position: relative;
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
}

.header__logo-container {
  margin: $s-size 0 0 $s-size;
  height: 4.8rem;
  width: 7rem;
}

.header__logo {
  width: 70%;
  height: auto;
}

.header__menu-container {
  height: 100%;
  list-style: none;
  /* margin: 0 20px 0 0; */
  display: flex;
  align-items: center;
  justify-content: space-around;
  font-size: $m-size;
  font-weight: 700;
}

.header__menu-item {
  border-bottom: 0.3rem solid transparent;
  margin: 0 $s-size;
  height: 100%;
}

.header__menu-item:hover {
  border-bottom: 0.3rem solid $dark-grey;
}

.header__menu-link {
  display: flex;
  align-items: center;
  height: 100%;
  text-decoration: none;
  color: $dark-grey;
}

.header__menu-button {
  height: 100%;
  border: none;
  background: none;
  color: $dark-grey;
  font-weight: 700;
  cursor: pointer;
  outline: none;
}

.header__modal {
  position: absolute;
  top: 50%;
  left: 50%;
  right: auto;
  bottom: auto;
  //margin-right: -50%;
  transform: translate(-50%, -50%);
  background: white;
  outline: none;
  padding: $m-size $l-size $l-size $l-size;
  width: 100%;
  max-width: 55rem;
}

.header__modal-overlay {
  position: fixed;
  top: 0px;
  left: 0px;
  right: 0px;
  bottom: 0px;
  background-color: rgba(0, 0, 0, 0.3);;
}

.header__title {
  font-size: 4vh;
  color: white;
  position: absolute;
  top: calc(100% + 22vh);
}

.header__close-button-container {
  overflow: hidden;
}

.header__close-button {
  border: none;
  background: none;
  outline: none;
  cursor: pointer;
  font-size: $l-size;
  padding-left: 0.3rem;
  padding-right: 0.3rem;
  color: $dark-grey;
  float: right;

}

@media (min-width: $desktop-breakpoint) {
  .header__title {
    position: relative;
    color: $dark-grey;
    font-size: 2.9rem;
    top: 0;
  }
  
  .header__menu-container {
    margin-right: $xs-size;
  }

  .header__menu-item {
    margin: 0 $m-size;
  }

  .header__logo-container {
    margin: $xm-size 0 0 $s-size;
    height: 5.6rem;
    width: 7.5rem;
  }
}

@media (max-width: $desktop-breakpoint) {
  .header__title {
    width: 100%;
    text-align: center;
    left: 50%;
    transform: translate(-50%, 0);
  }
}
