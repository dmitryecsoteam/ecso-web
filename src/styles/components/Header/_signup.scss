.signup__input {
    border: none;
    color: $light-black;
    width: 100%;
}

.signup__input:focus {
    outline: none;
}

.signup__title {
    color: $light-black;
    text-align: center;
    padding-bottom: $s-size;
}

@media(max-width: $desktop-breakpoint) {
    .signup__title {
        padding-top: $xs-size;
    }
}