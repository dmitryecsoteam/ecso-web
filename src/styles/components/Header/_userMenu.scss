.user-menu__image-container {
  border: 2px solid $dark-grey;
  border-radius: 50%;
  overflow: hidden;
  max-height: 3.5rem;
  max-width: 3.5rem;
}

.user-menu__image {
  height: 100%;
  width: 100%;
  position: relative;
  top: 0.2rem;
}

.user-menu__dropdown-container {
  position: relative;
  top: 2.1rem;
}

.user-menu__dropdown {
  position: absolute;
  right: 0;
  background-color: white;
  padding: $m-size;
  z-index: 10;
  cursor: default;
  box-shadow: 0 $xs-size $l-size -$m-size rgba(black, 0.3);
  animation: dropdown .2s ease-out forwards;
}

@keyframes dropdown {
  from {
    top: -3.2rem;
    opacity: 0;
  }
  to {
    top: 0;
    opacity: 1;
  }
}

.user-menu__name {
  color: $light-black;
  text-align: right;
  font-size: $xm-size;
}

.user-menu__email {
  text-align: right;
  font-size: $s-size;
  font-weight: 300;
  color: $mid-grey;
}

.user-menu__list {
  list-style: none;
  margin-top: $s-size;
}

.user-menu__item {
  white-space: nowrap;
}

.user-menu__item--top-border {
  border-top: 0.1rem solid black;
  padding-top: $xs-size;
}

.user-menu__link {
  text-decoration: none;
  color: $dark-grey;
  padding: $xs-size;
  display: block;
  cursor: pointer;
  font-weight: normal;
}

.user-menu__link:hover {
  color: $option-choose-menu;
  text-decoration: underline;
}

@media (max-width: $desktop-breakpoint) {
  .user-menu__dropdown-container {
    top: 1.4rem;
  }
}
