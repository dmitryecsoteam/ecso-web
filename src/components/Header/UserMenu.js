import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';

import userIcon from '../../images/icons/user.png';
import Signout from './Signout';



export default ({ user }) => {

    const [isActive, setIsActive] = useState(false);

    // Add global event listeners to catch clicks outside user menu to close it
    useEffect(() => {
        document.addEventListener("mousedown", handleClickOutside);
        return () => {
            document.removeEventListener("mousedown", handleClickOutside);
        }
    });

    const handleClickOutside = (event) => {
        
        // Don't change state if one of user menu items were clicked
        if (isActive && !event.target.className.includes('user-menu__link')) setIsActive(false);
    }

    const DropdownMenu = <div className="user-menu__dropdown-container">
        <div className="user-menu__dropdown">
            <h4 className="user-menu__name">{user.name}</h4>
            <p className="user-menu__email">{user.email}</p>
            <ul className="user-menu__list">
                <li className="user-menu__item user-menu__item--top-border">
                    <Link
                        className="user-menu__link"
                        to="/notifications"
                    >
                        My notifications
            </Link>
                </li>
                <li className="user-menu__item">
                    <Signout />
                </li>
            </ul>
        </div>
    </div>



    return (
        <button 
            className="header__menu-button" 
            onClick={() => setIsActive(!isActive)}
        >
            <div className="user-menu__image-container" >
                <img className="user-menu__image" src={userIcon} />
            </div>

            {isActive && DropdownMenu}

        </button>
    );
}

