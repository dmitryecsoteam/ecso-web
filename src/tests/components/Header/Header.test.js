import React from 'react';
import { shallow } from 'enzyme';
import Header from '../../../components/Header';

test('should render Header without title prop', () => {
    const wrapper = shallow(<Header />);
    expect(wrapper).toMatchSnapshot();
});

test('should render Header with title prop', () => {
    const wrapper = shallow(<Header title="Moscow, Russia" />);
    expect(wrapper).toMatchSnapshot();
});