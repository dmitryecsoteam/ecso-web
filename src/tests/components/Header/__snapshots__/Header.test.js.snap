// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header with title prop 1`] = `
<header
  className="header"
>
  <div
    className="header__logo-container"
  >
    <Link
      replace={false}
      to="/"
    >
      <img
        alt="logo"
        className="header__logo"
        src="test-file-stub"
      />
    </Link>
  </div>
  <h1
    className="header__title"
  >
    Moscow, Russia
  </h1>
  <nav>
    <ul
      className="header__menu-container"
    >
      <li
        className="header__menu-item"
      >
        <Link
          className="header__menu-link"
          replace={false}
          to="/help"
        >
          Help
        </Link>
      </li>
      <li
        className="header__menu-item"
      >
        <button
          className="header__menu-button"
          onClick={[Function]}
        >
          Signin
        </button>
      </li>
      <li
        className="header__menu-item"
      >
        <button
          className="header__menu-button"
          onClick={[Function]}
        >
          Signup
        </button>
      </li>
    </ul>
  </nav>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="header__modal"
    closeTimeoutMS={0}
    isOpen={false}
    onRequestClose={[Function]}
    overlayClassName="header__modal-overlay"
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
  >
    <div
      className="header__close-button-container"
    >
      <button
        className="header__close-button"
        onClick={[Function]}
      >
        ✕
      </button>
    </div>
    <Signup />
  </Modal>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="header__modal"
    closeTimeoutMS={0}
    isOpen={false}
    onRequestClose={[Function]}
    overlayClassName="header__modal-overlay"
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
  >
    <div
      className="header__close-button-container"
    >
      <button
        className="header__close-button"
        onClick={[Function]}
      >
        ✕
      </button>
    </div>
    <Signin />
  </Modal>
</header>
`;

exports[`should render Header without title prop 1`] = `
<header
  className="header"
>
  <div
    className="header__logo-container"
  >
    <Link
      replace={false}
      to="/"
    >
      <img
        alt="logo"
        className="header__logo"
        src="test-file-stub"
      />
    </Link>
  </div>
  <nav>
    <ul
      className="header__menu-container"
    >
      <li
        className="header__menu-item"
      >
        <Link
          className="header__menu-link"
          replace={false}
          to="/help"
        >
          Help
        </Link>
      </li>
      <li
        className="header__menu-item"
      >
        <button
          className="header__menu-button"
          onClick={[Function]}
        >
          Signin
        </button>
      </li>
      <li
        className="header__menu-item"
      >
        <button
          className="header__menu-button"
          onClick={[Function]}
        >
          Signup
        </button>
      </li>
    </ul>
  </nav>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="header__modal"
    closeTimeoutMS={0}
    isOpen={false}
    onRequestClose={[Function]}
    overlayClassName="header__modal-overlay"
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
  >
    <div
      className="header__close-button-container"
    >
      <button
        className="header__close-button"
        onClick={[Function]}
      >
        ✕
      </button>
    </div>
    <Signup />
  </Modal>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="header__modal"
    closeTimeoutMS={0}
    isOpen={false}
    onRequestClose={[Function]}
    overlayClassName="header__modal-overlay"
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
  >
    <div
      className="header__close-button-container"
    >
      <button
        className="header__close-button"
        onClick={[Function]}
      >
        ✕
      </button>
    </div>
    <Signin />
  </Modal>
</header>
`;
